### PR TITLE
[DOP-20059] Add granularity for `/runs/lineage`

### PR DESCRIPTION
**Pull request by TiGrib**
## Change Summary

Add granularity logic for `LineageService.get_lineage_by_runs()`

## Related issue number

[DOP-20059]

## Checklist

* [ ] Commit message and PR title is comprehensive
* [ ] Keep the change as small as possible
* [ ] Unit and integration tests for the changes exist
* [ ] Tests pass on CI and coverage does not decrease
* [ ] Documentation reflects the changes where applicable
* [ ] `docs/changelog/next_release/<pull request or issue id>.<change type>.rst` file added describing change
  (see [CONTRIBUTING.rst](https://github.com/MobileTeleSystems/data-rentgen/blob/develop/CONTRIBUTING.rst) for details.)
* [ ] My PR is ready to review.

### CODE REVIEW

File: data_rentgen/server/schemas/v1/lineage.py [hunk 1]
```diff
@@ -80,27 +83,50 @@ def _check_until(cls, value: datetime | None, info: ValidationInfo) -> datetime
...
+    )
 
 
 class OperationLineageQueryV1(BaseLineageQueryV1):
     start_node_id: UUID = Field(description="Operation id", examples=["00000000-0000-0000-0000-000000000000"])
+    granularity: Literal["OPERATION"] = Field(
+        description="Granularity of the operation lineage",
+        default="OPERATION",
+        examples=["OPERATION"],
+    )
```
Comment (dolfinus) [inline data_rentgen/server/schemas/v1/lineage.py:111]: Is there any need in such query param?

File: data_rentgen/server/schemas/v1/lineage.py [hunk 2]
```diff
@@ -80,27 +83,50 @@ def _check_until(cls, value: datetime | None, info: ValidationInfo) -> datetime
...
 
 
 class JobLineageQueryV1(BaseLineageQueryV1):
     start_node_id: int = Field(description="Job id", examples=[42])
+    granularity: Literal["JOB", "RUN", "OPERATION"] = Field(
+        description="Granularity of the job lineage",
+        # TODO: Change default to job in DOP-20060
+        default="OPERATION",
+        examples=["JOB", "RUN", "OPERATION"],
+    )
```
Comment (dolfinus) [inline data_rentgen/server/schemas/v1/lineage.py:102]: Could this be implemented in a dedicated PR?

File: pyproject.toml [hunk 3]
```diff
@@ -126,6 +126,7 @@ wemake-python-styleguide = "^0.19.2"
 flake8-pyproject = "^1.2.3"
 types-pyyaml = "^6.0.12"
 types-python-dateutil = "^2.9.0"
+icecream = "^2.1.3"
```
Comment (dolfinus) [inline pyproject.toml:129]: Not used in source code

File: tests/test_server/test_lineage/test_run_lineage.py [hunk 4]
```diff
@@ -872,3 +877,28 @@ async def test_get_run_lineage_with_symlinks(
...
+    response = await test_client.get(
+        "v1/runs/lineage",
+        params={
+            "since": run.created_at.isoformat(),
+            "start_node_id": str(run.id),
+            "direction": "DOWNSTREAM",
+        },
+    )
+
+    assert response.status_code == HTTPStatus.OK, response.json()
```
Comment (dolfinus) [inline tests/test_server/test_lineage/test_run_lineage.py:902]: The most interesting part is missing

File: data_rentgen/db/repositories/output.py [hunk 5]
```diff
@@ -83,6 +162,66 @@ async def list_by_dataset_ids(
         result = await self._session.scalars(query)
         return list(result.all())
 
+    async def list_by_dataset_ids_grouped_by_run(
```
Comment (dolfinus) [inline data_rentgen/db/repositories/output.py:165]: What if instead of adding 3 copies of the same method, we'll add an optional `group_by` method? Examples:
```python
dataset.list_by_dataset_ids([...], group_by=Operation)
dataset.list_by_dataset_ids([...], group_by=Run)
# or
dataset.list_by_dataset_ids([...], group_by=[Dataset.operation_id, Dataset.run_id])
dataset.list_by_dataset_ids([...], group_by=[Dataset.run_id])
```

File: pyproject.toml (hunk 3 above)
Comment (TiGrib) [reply pyproject.toml:129]: Isn't it would be helpful for debuging in dev environment ?

File: data_rentgen/db/repositories/output.py (hunk 5 above)
Comment (TiGrib) [reply data_rentgen/db/repositories/output.py:165]: I am not sure about unpacking result for this methods.
One row is tuple and we need to unpack it to create Input/Output instance. But for different params we would return different nums of elements in row:
For example:

```python
output.list_by_dataset_ids([...], group_by=Run) ->
Output(
                created_at=None,
                id=None,
                operation_id=None,
                job_id=None,
                run_id=row[0],
                dataset_id=row[1],
                num_bytes=row[2],
                num_rows=row[3],
                num_files=row[4],
            )

output.list_by_dataset_ids([...], group_by=Job) ->

Output(
                created_at=None,
                id=None,
                operation_id=None,
                job_id=row[0],
                run_id=None,
                dataset_id=row[1],
                num_bytes=row[2],
                num_rows=row[3],
                num_files=row[4],
            )

```

Comment (dolfinus) [reply data_rentgen/db/repositories/output.py:165]:
```python
output.list_by_dataset_ids([...], group_by=Run):
output.list_by_dataset_ids([...], group_by=Job):

query_group_by = [Dataset.id]
if group_by == Operation:
  query_group_by.append(Dataset.operation_id)
if group_by == Run:
  query_group_by.append(Dataset.run_id)
if group_by == Job:
  query_group_by.append(Dataset.job_id)

query = (
  select(
    Output.id if group_by == Operation else literal(None).alias("id"),
    Output.operation_id if group_by == Operation else literal(None).alias("operation_id"),
    Output.run_id if group_by in (Operation, Run) else literal(None).alias("run_id"),
    Output.job_id,
    Output.dataset_id,
    sum(Output.num_bytes).alias("num_bytes"),
    sum(Output.num_files).alias("num_files"),
    sum(Output.num_rows).alias("num_rows"),
  )
  .where(...)
  .group_by(*query_group_by)
```

And then convert Row to Output

File: pyproject.toml (hunk 3 above)
Comment (dolfinus) [reply pyproject.toml:129]: Ok

File: data_rentgen/server/services/lineage.py [hunk 6]
```diff
@@ -194,20 +203,31 @@ async def get_lineage_by_runs(
...
+                else:
+                    inputs = await self._uow.input.list_by_operation_ids_grouped_by_run(sorted(operation_ids))
 
         # Return only operations which have at least one input or output
-        operation_ids = {input.operation_id for input in inputs} | {output.operation_id for output in outputs}
+        # In case granularity == "RUN" operations will be empty
+        ids_to_skip = ids_to_skip or IdsToSkip()
+        operation_ids = {input.operation_id for input in inputs} | {
+            output.operation_id for output in outputs
+        } - ids_to_skip.operations
```
Comment (dolfinus) [inline data_rentgen/server/services/lineage.py:226]: Probably we should exclude items with `operation_id == null`

File: data_rentgen/server/services/lineage.py [hunk 7]
```diff
@@ -194,20 +203,31 @@ async def get_lineage_by_runs(
 
         all_operations = await self._uow.operation.list_by_run_ids(sorted(runs_by_id.keys()), since, until)
         operation_ids = {operation.id for operation in all_operations}
```
Comment (dolfinus) [inline data_rentgen/server/services/lineage.py:205]: You can move this under `case "OPERATION":`

File: tests/test_server/test_lineage/test_run_lineage.py [hunk 8]
```diff
@@ -721,24 +821,394 @@ async def test_get_run_lineage_with_depth_ignore_cycles(
...
+                "started_by_user": {"name": run.started_by_user.name},
+                "start_reason": run.start_reason.value,
+                "ended_at": run.ended_at.strftime("%Y-%m-%dT%H:%M:%SZ"),
+                "end_reason": run.end_reason,
+            },
+        ],
+    }
+
+
+async def test_get_run_lineage_with_depth_ignore_cycles_with_operation_granularity(
```
Comment (dolfinus) [inline tests/test_server/test_lineage/test_run_lineage.py:932]: Do we really need this test?

File: tests/test_server/test_lineage/test_run_lineage.py [hunk 9]
```diff
@@ -872,3 +1343,115 @@ async def test_get_run_lineage_with_symlinks(
             for operation in sorted(operations, key=lambda x: x.id)
         ],
     }
+
+
+async def test_get_run_lineage_with_operation_granularity(
```
Comment (dolfinus) [inline tests/test_server/test_lineage/test_run_lineage.py:1348]: Let's move this test before `with_depth`

File: data_rentgen/db/repositories/input.py [hunk 10]
```diff
@@ -75,11 +75,87 @@ async def list_by_operation_ids(
...
+            Input.run_id,
+            Input.job_id,
+            Input.dataset_id,
+            Input.num_bytes,
+            Input.num_rows,
+            Input.num_files,
+        ).where(
+            Input.created_at >= min_created_at,
+            Input.run_id == any_(run_ids),  # type: ignore[arg-type]
+        )
```
Comment (dolfinus) [inline data_rentgen/db/repositories/input.py:98]: Hm, but where is a `group_by` clause?

File: data_rentgen/db/repositories/input.py [hunk 11]
```diff
@@ -134,3 +246,22 @@ async def _update(self, existing: Input, new: InputDTO) -> Input:
             existing.num_files = new.num_files
         await self._session.flush([existing])
         return existing
+
+    async def _list_with_aggregation(self, filters: list[ColumnElement[bool]], aggregation_field: str):
```
Comment (dolfinus) [inline data_rentgen/db/repositories/input.py:250]: This method is not used anywhere

File: data_rentgen/server/api/v1/router/dataset.py [hunk 12]
```diff
@@ -48,13 +48,14 @@ async def search_datasets(
...
 @router.get("/lineage", summary="Get Dataset lineage graph")
-async def get_dataset_lineage(
+async def get_datasets_lineage(
     pagination_args: Annotated[DatasetLineageQueryV1, Query()],
     lineage_service: Annotated[LineageService, Depends()],
 ) -> LineageResponseV1:
     lineage = await lineage_service.get_lineage_by_datasets(
         start_node_ids=[pagination_args.start_node_id],  # type: ignore[list-item]
         direction=pagination_args.direction,
+        granularity=pagination_args.granularity,
```
Comment (dolfinus) [inline data_rentgen/server/api/v1/router/dataset.py:58]:
```suggestion
        granularity="OPERATION",
```

File: pyproject.toml [hunk 13]
```diff
@@ -159,6 +160,7 @@ markers = [
     "server: tests for server (require running database)",
     "consumer: tests for consumer (require running database)",
     "db: tests for db",
+    "lineage: tests for lineage endpoints (require running database)"
```
Comment (dolfinus) [inline pyproject.toml:163]: nit: you can actually use `pytest -k test_lineage`, as all lineage tests are placed in the same directory with this exact name

**Review by dolfinus — APPROVED** [review]
